tests/protocol/test_protocol.py

```diff
@@ -136,6 +136,28 @@ async def test_client_files_stat(client):
     assert len(result.result["files"]) == len(other_files)
 
 
+async def test_upload_file_twice(client):
+    """
+    This test checks that attempts to re-upload the same file will be silently ignored.
+    """
+
+    async def upload_and_stat(body: str, client, hash: str):
+        result = await client.upload_file(id=hash, content=body)
+        assert result.code == 200
+        result = await client.stat_files(files=[hash])
+        assert result.code == 200
+        assert len(result.result["files"]) == 0
+
+    (hash, content, body) = make_random_file()
+
+    result = await client.stat_files(files=[hash])
+    assert result.code == 200
+    assert len(result.result["files"]) == 1
+
+    await upload_and_stat(body, client, hash)
+    await upload_and_stat(body, client, hash)
```

Comment:
To test whether the concurrency issue is resolved, you should run these two API calls concurrently.
```suggestion
    await asyncio.gather(upload_and_stat(body, client, hash),  upload_and_stat(body, client, hash))
```